File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let result = 0;
    let cakeA = {};
    let cakeACount = 0; // cakeA의 토핑의 수
    let cakeB = new Set();
    
    // cakeA에 모든 토핑을 옮김
    for(let i = 0; i < topping.length; i++) {
        if (cakeA[topping[i]]) {
            cakeA[topping[i]]++;
        } else {
            cakeA[topping[i]] = 1;
            cakeACount++;
        }
    }
    
    // 하나씩 cakeB로 옮기면서 토핑의 개수 확인
    for(let i = 0; i < topping.length; i++) {
        cakeB.add(topping[i]);
        cakeA[topping[i]]--;
        
        if (!cakeA[topping[i]]) cakeACount--;
        
        if (cakeACount === cakeB.size) result++;
    }
    
    return result;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    result = 0
    me = Counter(topping)
    bro = set()
    
    for t in topping:
        me[t] -= 1
        if me[t] == 0: me.pop(t)
        bro.add(t)
        if len(me) == len(bro):
            result += 1
    
    return result
        